### PR TITLE
fix to prioritize newly added registry

### LINKED ISSUE #6283: Newly added registry no longer has priority with `odo registry` and `odo init`
/kind bug

This is an issue following up on the merge of https://github.com/redhat-developer/odo/pull/6266

---
Now that the Preference Client implementation does the reverse operation (which is a good thing), I think there is one side effect, which we should consider, IMO.
In [`pkg/registry/registry.go`](https://github.com/redhat-developer/odo/blob/main/pkg/registry/registry.go#L64-L65), the registry list is also iterated from the end, which again returns the list in the wrong order, for example when running `odo registry` or even when fetching a new Devfile with `odo init` :)

<details>
<summary>Examples</summary>

```sh
# Correct order: StagingRegistry comes first. Same with JSON output.
$ odo preference view

...
Devfile registries:
 NAME                    URL                                SECURE 
 StagingRegistry         https://registry.stage.devfile.io  No     
 DefaultDevfileRegistry  https://registry.devfile.io        No     
```

```sh
# Issue 1: For each Devfile name, StagingRegistry should be listed first
$ odo registry

...
NAME                          REGISTRY                DESCRIPTION                                 
 dotnet50                      DefaultDevfileRegistry  Stack with .NET 5.0                         
 dotnet50                      StagingRegistry         Stack with .NET 5.0                         
 dotnet60                      DefaultDevfileRegistry  Stack with .NET 6.0                         
 dotnet60                      StagingRegistry         Stack with .NET 6.0                         
 dotnetcore31                  DefaultDevfileRegistry  Stack with .NET Core 3.1                    
 dotnetcore31                  StagingRegistry         Stack with .NET Core 3.1                    
 go                            DefaultDevfileRegistry  Go is an open source programming languag... 
 go                            StagingRegistry         Go is an open source programming language... 
```

```sh
# Issue 2: Devfile should have been selected from "StagingRegistry"
$ odo init           
  __
 /  \__     Initializing a new component
 \__/  \    Files: Source code detected, a Devfile will be determined based upon source code autodetection
 /  \__/    odo version: v3.2.0
 \__/

Interactive mode enabled, please answer the following questions:
Based on the files in the current directory odo detected
Language: Java
Project type: Maven
The devfile "java-maven" from the registry "DefaultDevfileRegistry" will be downloaded.
? Is this correct? (Y/n) 
```

</details>

So I think we need to change [`pkg/registry/registry.go`](https://github.com/redhat-developer/odo/blob/main/pkg/registry/registry.go#L64-L65) to rely on the order returned by the Preference Client.

And we should also make sure that there are no other places where the list returned by the Preference Client is iterated in the reverse order ^^

_Originally posted by @rm3l in https://github.com/redhat-developer/odo/pull/6266#discussion_r1013707569_

### PR DESCRIPTION
**Pull request by anandrkskd**
Signed-off-by: anandrkskd <anandrkskd@gmail.com>

**What type of PR is this:**

/kind bug
**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #6283 

**PR acceptance criteria:**

- [x] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
```
$ odo preference add registry StageDevfileRegistry https://registry.stage.devfile.io

$ odo preference view                                                               
Preference parameters:
 PARAMETER           VALUE           
 ConsentTelemetry    false (default) 
 Ephemeral                           
 PushTimeout                         
 RegistryCacheTime                   
 Timeout                             
 UpdateNotification                  

Devfile registries:
 NAME                    URL                                SECURE 
 StageDevfileRegistry    https://registry.stage.devfile.io  No     
 DefaultDevfileRegistry  https://registry.devfile.io        No    
 
$ odo registry        
 NAME                          REGISTRY                DESCRIPTION                                 
 dotnet50                      StageDevfileRegistry    Stack with .NET 5.0                         
 dotnet50                      DefaultDevfileRegistry  Stack with .NET 5.0                         
 dotnet60                      StageDevfileRegistry    Stack with .NET 6.0                         
 dotnet60                      DefaultDevfileRegistry  Stack with .NET 6.0                         
 dotnetcore31                  StageDevfileRegistry    Stack with .NET Core 3.1                    
 dotnetcore31                  DefaultDevfileRegistry  Stack with .NET Core 3.1     
```

### CODE REVIEW

**Comment by dharmit** [comment]
Adding both flags since it's a small fix.

/lgtm
/approve

Adding hold so that failures don't cause the tests to run in an infinite loop. Remove hold when tests are fixed.

/hold

**Comment by rm3l** [comment]
@anandrkskd I think there is still a side effect with the order (still due to https://github.com/redhat-developer/odo/pull/6266), when we add more than 2 registries.

```shell
$ export GLOBALODOCONFIG=/tmp/my-odo-preferences.yaml

$ odo preference view
...     

Devfile registries:
 NAME                    URL                          SECURE 
 DefaultDevfileRegistry  https://registry.devfile.io  No     

$ odo preference add registry StagingRegistry https://registry.stage.devfile.io
New registry successfully added

$ odo preference add registry MyFakeRegistry https://my-fake-registry.example.com
New registry successfully added

# Order returned by `odo preference view` is not correct, I think.
# Order should be: MyFakeRegistry, StagingRegistry, DefaultDevfileRegistry
$ odo preference view                                                            
..
Devfile registries:
 NAME                    URL                                   SECURE 
 MyFakeRegistry          https://my-fake-registry.example.com  No     
 DefaultDevfileRegistry  https://registry.devfile.io           No     
 StagingRegistry         https://registry.stage.devfile.io     No     

# The order in which the registries are stored in the preferences is not correct either.
# Order in this file should be: DefaultDevfileRegistry, StagingRegistry, MyFakeRegistry
$ cat /tmp/my-odo-preferences.yaml
kind: Preference
apiversion: odo.dev/v1alpha1
OdoSettings:
  RegistryList:
  - Name: StagingRegistry
    URL: https://registry.stage.devfile.io
    secure: false
  - Name: DefaultDevfileRegistry
    URL: https://registry.devfile.io
    secure: false
  - Name: MyFakeRegistry
    URL: https://my-fake-registry.example.com
    secure: false
  ConsentTelemetry: true
```

This issue was due to https://github.com/redhat-developer/odo/pull/6266.
I think it is a good thing that the preference client does the reverse ordering, but maybe we need to reverse the order when adding a new registry.

File: pkg/odo/cli/preference/view.go [hunk 1]
```diff
@@ -85,8 +85,8 @@ func HumanReadableOutput(preferenceList preference.PreferenceList, registryList
 	registryT.AppendHeader(table.Row{"NAME", "URL", "SECURE"})
 
 	// Loop backwards here to ensure the registry display order is correct (display latest newly added registry firstly)
```
Comment (rm3l) [inline pkg/odo/cli/preference/view.go:87]: This comment is no longer relevant and can be removed.
```suggestion
```

File: pkg/odo/cli/preference/view.go [hunk 2]
```diff
@@ -85,8 +85,8 @@ func HumanReadableOutput(preferenceList preference.PreferenceList, registryList
...
-	for i := range *registryList {
-		registry := (*registryList)[i]
+	for i := range registryList {
+		registry := (registryList)[i]
```
Comment (rm3l) [inline pkg/odo/cli/preference/view.go:89]: nit: 
```suggestion
		registry := registryList[i]
```

File: pkg/preference/implem.go [hunk 3]
```diff
@@ -437,12 +424,21 @@ func (c *preferenceInfo) ConsentTelemetry() *bool {
 	return c.OdoSettings.ConsentTelemetry
 }
 
-func (c *preferenceInfo) RegistryList() *[]Registry {
-	return c.OdoSettings.RegistryList
+func (c *preferenceInfo) RegistryList() []Registry {
```
Comment (rm3l) [inline pkg/preference/implem.go:427]: Can you document this method, explaining why the order is being reversed?

File: pkg/registry/registry.go [hunk 4]
```diff
@@ -60,10 +60,9 @@ func (o RegistryClient) GetDevfileRegistries(registryName string) ([]api.Registr
 
 	hasName := len(registryName) != 0
 	if o.preferenceClient.RegistryList() != nil {
-		registryList := *o.preferenceClient.RegistryList()
+		registryList := o.preferenceClient.RegistryList()
 		// Loop backwards here to ensure the registry display order is correct (display latest newly added registry firstly)
```
Comment (rm3l) [inline pkg/registry/registry.go:64]: Same point about removing this comment, which is no longer relevant here.

**Review by rm3l — COMMENTED** [review]
LGTM feature-wise. Thanks for fixing the behavior.

Added a few comments.

File: pkg/preference/implem.go (hunk 3 above)
Comment (rm3l) [reply pkg/preference/implem.go:430]: A suggestion to fix the wording and a few typos:

```go
// RegistryList returns the list of registries declared in the preferences file.
// Adding a new registry always adds it to the end of the list in the preferences file, but RegistryList intentionally reverses the order to prioritize the most recently added registries.
func (c *preferenceInfo) RegistryList() []Registry {
```

**Pull request merged**